app.rb

```diff
@@ -18,194 +18,288 @@
-      if data.keys.include?('___checkbox_unchecked')
-        data.keys.include?('___checkbox_checked')
-      elsif data.keys.all? { |k| k =~ /\A\d+\Z/ }
-        data.to_a.sort_by(&:first).map(&:last).map &method(:normalize)
-      else
-        data.merge(data) { |k,v| normalize(v) }
-      end
-    else
-      data
+  class File
+    def initialize(path, name=nil)
+      @path = path
+      @name = name
     end
-  end
 
-  helpers do
-    def render_yml(value)
-      case value
-      when Hash
-        value.map { |k, v| render_value(k, v, "data[#{k}]") }.join
-      when Array
-        value.each_with_index.map { |v, i| render_value(nil, v, "data[#{i}]") }.join
-      else
-        raise TypeError, 'YAML root must be a Hash or Array'
-      end
+    def directory?
+      ::File.directory?(path)
     end
 
-    def render_value(key, value, name)
-      case value
-      when Hash
-        "<fieldset>#{"<legend>#{key}</legend>" if key}#{
-        value.map do |key, value|
-          "#{render_value(key, value, "#{name}[#{key}]")}"
-        end.join
-        }</fieldset>"
-      when Array
-        "<fieldset><legend>#{key}</legend><ol>#{
-          value.each_with_index.map { |v, i| "<li>#{render_value(nil, v, "#{name}[#{i}]")}</li>" }.join
-        }</ol></fieldset>"
-      else
-        render_input(key, value, name)
+    def text?
+      FileMagic.open(:mime) do |magic|
+        magic.file(path).split('/').first == 'text'
       end
     end
 
-    def render_input(key, value, name)
-      input = if [true, false].include? value
-        "
-          <input type='hidden' name='#{name}[___checkbox_unchecked]' value='' />
-          <input type='checkbox' name='#{name}[___checkbox_checked]' #{' checked="checked"' if value} />
-        "
-      elsif value.is_a?(String) && value.length > 50
-        "<textarea name='#{name}'>#{value}</textarea>"
-      else
-        "<input type='text' name='#{name}' value='#{value}' />"
+    def markdown?
+      ['md', 'markdown'].include? extension
+    end
+
+    def yml?
+      ["yaml", "yml"].include? extension
+    end
+
+    def root?
+      path == PROJECT_PATH
+    end
+
+    def route
+      "/files#{path}".gsub(PROJECT_PATH, "")
+    end
+
+    def name
+      @name ||= ::File.basename(path)
+    end
+
+    def stream_content
+      has_frontmatter = false
+      doc_start = 0
+      ::File.foreach(path).with_index do |line, idx|
+        line.chomp!
+
+        if doc_start < 2
+          has_frontmatter = true if line == '---' && idx == 0
+          doc_start += 1 if line == '---'
+          next if has_frontmatter
+        end
+
+        yield line
       end
+    end
 
-      if key
-        "<label>#{key} #{input}</label>"
-      else
-        input
+    def content?
+      detect_content_or_frontmatter[0]
+    end
+
+    def frontmatter?
+      detect_content_or_frontmatter[1]
+    end
+
+    def frontmatter
+      YAML.safe_load(open(path))
+    end
+
+  protected
+
+    def extension
```

Comment:
Right now, since I'm not using it outside the class and have the two wrapper methods I decided to keep it protected, but I'm not opposed to making it public in the future if there was a need for that.